add flag for turn shutdown complete. At turnConnectionShutdown, run c…

chehefen opened the pull request:
…leanup regardless of having allocation or not to clean up the socket. Add unit test covering turn socket clean up

*Issue #, if available:*
fix #649 

*Description of changes:*


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

MushMal commented on src/source/Ice/ConnectionListener.c:
```diff
@@ -129,6 +131,19 @@ STATUS connectionListenerRemoveConnection(PConnectionListener pConnectionListene
+    while (!hasConnection && pCurNode != NULL) {
+        pCurrSocketConnection = (PSocketConnection) pCurNode->data;
+        pCurNode = pCurNode->pNext;
+        if (pCurrSocketConnection == pSocketConnection) {
+            hasConnection = TRUE;
+        }
```
Nitpick. hasConnection = (pCurrSocketConnection == pSocketConnection);

MushMal commented on src/source/Ice/TurnConnection.c:
```diff
@@ -211,6 +218,8 @@ STATUS turnConnectionHandleStun(PTurnConnection pTurnConnection, PBYTE pBuffer,
 
     switch (stunPacketType) {
         case STUN_PACKET_TYPE_ALLOCATE_SUCCESS_RESPONSE:
+            /* If shutdown has been initiated, ignore the alloaction response */
```
*allocation

MushMal commented on src/source/Ice/TurnConnection.c:
```diff
@@ -1048,10 +1057,9 @@ STATUS turnConnectionStepState(PTurnConnection pTurnConnection)
 
                 CHK_STATUS(turnConnectionFreePreAllocatedPackets(pTurnConnection));
-
+                CHK_STATUS(socketConnectionClosed(pTurnConnection->pControlChannel));
```
I am wondering if you still need to set the shutdownComplete atomic in case of failure anyway?

chehefen replied:
I added logic that when failure happens, turn will go to cleanup state, then go to failed state. Also added unit test for that.

chehefen commented:
> This looks good to me overall. Could you also add a unit test that closes one of the sockets from the connection list, and verify that it's deleted from the list properly by waiting on `pConnectionListener->removeConnectionComplete` condition variable?
In the two new unit tests I added, at the end I added check for turnSocketConnection to not exist in connectionListener's connection list.
